linux-capture: Fix Problems with the Window-Selection of the XComposite Source

MaZderMind opened the pull request:
### Description
This PR fixes the **three Problems** reported in #3040 that have been found in the Window-Source (XComposite) of the linux-capture Plugin:

TL;DR Video: https://mazdermind.de/dl/github/obs/3040/2.mp4


#### Problem 1
When a selected Window is closed, OBS randomly chooses another Window of the same Application (for example any other Browser-Window) if the initial window is closed. Because of that restoring the, maybe accidentally closed Window, will not restore it in the Scene.

The reason for this is, that if no Window of the selected ID can be found, a search based on the Window-Title or the Window-Class is started (See [Lines 242-244 of xcompcap-main.cpp](https://github.com/obsproject/obs-studio/blob/fd66869/plugins/linux-capture/xcompcap-main.cpp#L242-L244) for the quite convoluted comparison). If no Window of the same Title is found, any window of the same Class is selected. This Search-Branch that checks the Window-Class has been removed in this PR.

#### Problem 2
Cannot capture two windows with the same title: Tough the Selection Dialog lists both Windows, only one of them can be actually selected. Selecting the other one behaves as if the first one had been selected.

The reason for this is a programming flaw in `getWindowFromString` where in [Lines 239-240 of xcompcap-main.cpp](https://github.com/obsproject/obs-studio/blob/fd66869/plugins/linux-capture/xcompcap-main.cpp#L239-L240) `wid` is never assigned and stays `0` for the entire function. Therefore `cwin == wid` is always  `false` and the only relevant condition is `wname == cwinname && wcls == ccls` which only matched Windows by Name & Class and completely ignoring the selected Window-ID.

The relevant selection Loop has been replaced with two Loops: One tries to find the exact window by ID, the second Loop does a further search by Name & Class, if the exact Window has not been found.

The relevant Code is not in the hot Path and only runs every 2 Seconds while a Window is not open, so the increased readability of the two distince loops outweighs the possibly reduced performance from the nearly unreadable original loop (which has hidden the above bug for quite some time).

#### Problem 3
Event-Handling-Problems when capturing the same Window more then once. When creating two or more sources which each capture the same Window, the Capture works on first glance, but when modifying the Window (ie. by simply resizing it) one can see, that only one of the sources correctly responds. As a followup problem, this breaks the image-capturing of the other source, that did not correctly receive the events.

This can also be observed, when the Window is closed: One of the Sources gets the `DESTROY`-Notification, while the other does not get notified of the destruction and continues to grab from the non-existing window, resulting in X-Errors being printed to the console:
```
error: X Error: BadWindow (invalid Window parameter), Major opcode: BadWindow (invalid Window parameter), Minor opcode: 0, Serial: 5332
```

The Reason for this is, that `XCompcap::processEvents` stores changed Window-IDs in an `std::unordered_set` which then contains one entry per changed Window. A following call to `XCompcap::windowWasReconfigured` returns for that specific window returns `true` but also removes the Window-ID from the set, so that the second call from the other Source incorrectly returns `false`. (See [Lines 250-262 of xcompcap-helper.cpp](https://github.com/obsproject/obs-studio/blob/fd668695db7ed5724888e71b4df34285fd72b495/plugins/linux-capture/xcompcap-helper.cpp#L250-L262)).

~~After trying various approaches, I decided to completely move the State into the Source, so that now each Source has its own `Display`-Reference and receives only Events for its own Window. All Helper-Functions have been extended with an `Display *xdisplay` parameter which makes them only dependent from the respective Source. Also the `pthread_mutex_t changeLock` could be removed, which also removes a possible congestion in the hot path.~~

After @kkartaltepe found issues with this approach I removed the offending change and implemented a different solution, which now uses [a combination of `map` and `unordered_set`](https://github.com/MaZderMind/obs-studio/commit/05bc26797a3f167a506cd2255b2dd31279bca0b8#diff-52b602eca61b7f1851c824aea05e4f6fR169-R170) to track which sources are interested in Events from which Window and handles subscribing and unsubscribing in the shared code, thus correctly handling multiple sources capturing the same window.

I also had to fix [a Problem with xcc_cleanup](https://github.com/MaZderMind/obs-studio/commit/05bc26797a3f167a506cd2255b2dd31279bca0b8#diff-73ba83f076f2731c616a0940f79f585bL373-R377) cleaning up the Window's Compositor-Redirection for one of the two sources when a captured Window re-appeard after it was closed before.

### Motivation and Context
I recently produced Remote-Conference-Talks with multiple Speakers and a Moderator, which all were present in a _Google Meet_ (Google Video-Conferencing System). I had 4 Chrome-Browser windows open, which all participated in the Meet. Each of them had one Source pinned and was switched to Fullscreen.

With the Window-Capture I was able to select them as individual "Camera" Sources and create a nice mix. Because all Browser-Windows have the same Title in this Scenario I hit _Problem 2_, which could be worked around by changing the Window-Title via the Browser-Developer-Console.

_Problem 1_ occured when I first started OBS without configuring the Meet-Windows first. OBS would just select any open Browser-Window and keep that way, even after correctly configuring the Meet-Windows.

_Problem 3_ only occured in the Lab after I fixed _Problem 1 & 2_.

### How Has This Been Tested?
I tested these Changes on an `Ubuntu 19.10` installation under Lab conditions; see Video linked at the top.

### Types of changes
 - Bug fix (non-breaking change which fixes an issue)

Although that depends on whether the Logic to switch to another Window of the same Application when the original Window is closed county as Feature, but as it is totally uncontrollable which window the Source switches to, I rather see it as a fixed bug rather then functionality which has been removed.

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

MaZderMind commented:
Sorry for opening the PR incomplete. I hit the Keyboard Shortcut early. I'll complete the Text in a Second.

kkartaltepe commented:
Problem 3 only seems more aggravated with this change. Though some might say its an improvement that capture entirely fails over having a stale frame.

kkartaltepe commented:
Please make sure all of your commits are formatted and there are no cleanup commits like `dda9504` in the PR.

MaZderMind commented:
@kkartaltepe Re _Problem 3_: In current obs 25.0.8 and also on master, with two sources capturing the same window, resizing the window can make one of the sources fail. Closing the window leads to one source correctly cleaning up while the other logs error messages to the console (while as a side-effect of not being cleaned up properly) still shows a stalled frame. This clearly indicates a bug to me. See this Clip reproducing this Issue with the current OBS release: https://mazdermind.de/dl/github/obs/3040/3.mp4

Re the Problems you found capturing the same Window multiple times: This PR should indeed fix this scenario. Can you tell me the OS / Environment you get these Problems on? Maybe you can also create a Screen-Capture of your process?  I used [an ffmpeg-Commandline](https://p.rrbone.net/paste/ShMC+eQu#GsFtpNA7Qv+l48kquuSvd5MtBCvxs+movTR7bvsfcqn) to capture my two Screens while working with OBS.

I'll squash the Reformat-Commit into the individual commits as appropriate in a second. I'l like to first analyse and if possible fix the Issues that you reported

MaZderMind commented:
@kkartaltepe The `Did not find Window By ID %s, Name '%s' or Class '%s'` would be logged for ever Window that could not be found every 2 Seconds. Do you think this is too much for a configuration problem in the current scene? I'm not familar with the logging policy, but I can see that other Sources only log once when a Resource can not be found. I'll change that.

The `glXCreatePixmap` message is interesting; I'n not so deep into the glx functionality to know if there is anything special about capturing the same Window multiple times. It would be interesing if you could reproduce the same problem on `master` and on 75a79c048daf3fc2456f0271b9d3b4a15a02722a, which is before I moved the XDispay-Variable inside the Source. I'll try to reproduce the Problem here on different computers, too.

kkartaltepe commented:
>  Can you tell me the OS / Environment you get these Problems on? 

```
info: CPU Name: Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
info: CPU Speed: 3692.696MHz
info: Physical Cores: 4, Logical Cores: 8
info: Physical Memory: 32071MB Total, 5515MB Free
info: Kernel Version: Linux 5.6.13-arch1-1
info: Distribution: "Arch Linux" Unknown
info: Window System: X11.0, Vendor: The X.Org Foundation, Version: 1.20.8
info: Portable mode: false
info: Loading up OpenGL on adapter NVIDIA Corporation GeForce GTX 1070/PCIe/SSE2
info: OpenGL loaded successfully, version 3.3.0 NVIDIA 440.82, shading language 3.30 NVIDIA via Cg compiler
```
In an i3 session.
75a79c0 behaves the same as master.

> would be logged for every Window that could not be found every 2 Seconds.

Any logging of a steady state is undesirable.

MaZderMind commented:
@kkartaltepe I reduced the Logging-Level of that particular Message to `DEBUG` which by-default does not show up in the console. I'll try to reproduce your report in a similar environment. I so far was not able to reproduce it on my Workstation (AMD RX 460) or my Notebook (Intel DRI HD Graphics 3000). I don't think this is GPU related, probably a Problem with my usage of the X-Server API (although I don't think I have changed much in this respect)

MaZderMind commented:
I checked the two calls to `glXCreatePixmap` and `glXDestroyPixmap` and they are balanced, whatever I do to the windows; so every created pixmap is destroyed before a new one is created. The only difference between master and this PR is, that they do belong to different X-Display instances instead of a shared one.

Setting up an Arch/X11/i3 environment seems challenging. I would be happy if you could aide in debugging this issue which I can't reproduce on my two Ubuntu Machines.
 - Which Actions *exactly* do trigger the Problem?
 - Can you reproduce them booting from an Ubuntu Live-Stick?
 - Does the X-Server log anything?

MaZderMind commented:
I rebased the non-feature commits out of the PR by applying `formatcode.sh` to each commit individually.

kkartaltepe commented:
I cannot replicate on ubuntu on intel (gnome or i3). I will try to test on the nvidia stack if I can.

MaZderMind commented:
@kkartaltepe I'll try to come up with a different solution to the original Problem over the Weekend. Thank you again for testing this on your systems, as I can't reproduce the Problem locally.

MaZderMind commented:
@kkartaltepe I removed the offending change and implemented a different solution, which now uses [a combination of `map` and unordered_set`](https://github.com/MaZderMind/obs-studio/commit/05bc26797a3f167a506cd2255b2dd31279bca0b8#diff-52b602eca61b7f1851c824aea05e4f6fR169-R170) to track which sources are interested in Events from which Window and handles subscribing and unsubscribing in the shared code, thus correctly handling multiple sources capturing the same window.

I also had to fix [a Problem with xcc_cleanup](https://github.com/MaZderMind/obs-studio/commit/05bc26797a3f167a506cd2255b2dd31279bca0b8#diff-73ba83f076f2731c616a0940f79f585bL373-R377) cleaning up the Window's Compositor-Redirection for one of the two sources when a captured Window re-appeard after it was closed before.

kkartaltepe commented:
I think its sort of working as you describe. However it appears to be leaking X resources during events like resizing. If you resize a window for a minute or so OBS will crash (at least on ubuntu machine).

I havnt taken a look at any of the code yet, if I can make some time I will. Thanks for working on this.

MaZderMind commented:
@kkartaltepe I think I know where I messed up. I'll see to fix this over the weekend and let you know.

> Thanks for working on this.
Thynk oyu for checking and reviewing so thoroughly, that really helps.

MaZderMind commented:
@kkartaltepe I'm sorry that this PR takes so long. There is not much time between employment and family. I'm sure I reverted my mistake that caused the resource exhaustion but still fixed the problem where one source would un-redirect the window upon initialization, causing previously initialized sources to fail, when the matching window is started after obs.

seanlaguna commented:
It would be amazing to get this PR merged. I'm having issues where the naming of my windows is changing unpredictably, and sometimes OBS will actually change the window being captured without the window even opening or closing. The fact that it's completely unpredictable which window will appear is scary, meaning if I'm streaming I have to be certain that any window I have open isn't going to leak important information to my stream if it randomly gets switched to.

seanlaguna commented:
I actually just cloned this branch and it seems to have functionally the same problems that I had before. Maybe I should open a new issue?

From the original issue, I saw this:
> It should show some a Placeholder that the desired Window is not present

This seems like by far the most desirable behavior. I would like to contribute to this PR to help achieve this. Having an option to select a "placeholder image" for when the Window is not present would be ideal.

kkartaltepe commented:
Just getting back to reviewing things myself. I think the latest changes look pretty good to me for problems 1/2.

Im still trying to test 3 on intel. But on nvidia it consistently results in only the first capture working successfully. This seems better than the current behavior on nvidia where only one capture works at a time but during resizes the working capture may change and the 2nd always ends with a BadAlloc from glXCreatePixmap.

This change seems to cause BadAlloc on initial capture more consistently on i3/nvidia as well. Windows like firefox or some qt applications appear to get a BadAlloc on first capture but capture works after the window is resized (this is similar to an issue we see mostly on gnome, unsure if related).

There is still some log spam from processing messages.
`info: xcompcap: processEvents(): sourceChanged=0x55db32c6bb30`

kkartaltepe commented:
I cannot find any regressions on fedora 33 using gnome/intel.

kkartaltepe commented:
I went ahead and rebased and squashed and moved most of the logging to debug level. Should be mergable now.

jp9000 commented:
"Capture window by id first": Does this potentially break existing configurations?

kkartaltepe commented:
When someone starts obs there will be no id to capture so it will continue to use the last window class saved. Once someone actually captures a window with the new setup it allows them actually capture two different windows with the same class where as previously you could only capture one of those windows. And once a user closes one of those windows the behavior of falling back to another window in the same class is preserved.

Essentially the only change in behavior is that now users can capture two different windows of the same name.

MaZderMind commented:
@kkartaltepe Thank you very much for working through this.
